tests/cpp/test_multidevice_pipeline.cpp

```diff
@@ -576,6 +581,302 @@ INSTANTIATE_TEST_SUITE_P(
         SchedulingMode::Manual,
         SchedulingMode::ReductionOnly,
         SchedulingMode::Automatic));
+
+class DistributedMatmulTest : public MultiDeviceTest {
+ protected:
+  DistributedMatmulTest() : optimization_guard_(false) {
+    DisableOptionsGuard::getCurOptions().set(DisableOption::MatmulExprEval);
+  }
+
+  void SetUp() {
+    if (!deviceMajorMinorCheck(8)) {
+      GTEST_SKIP() << "Distributed matmul tests require Ampere or newer";
+    }
+    MultiDeviceTest::SetUp();
+    num_devices = communicator->size();
+  }
+
+  // use_fusion_executor cache=true, skip_auto_scheduling=false
+  // cache_fusion_executor=false
+  MultiDeviceExecutorParams executor_params{true, false, false};
+  int num_devices;
+
+  ValidationConstants getTolerances() {
```

Comment:
It's a bug if we have to manually tune these tolerances. The auto-computed threshold by testValidate should have already considered these errors caused by reduction order. I don't see why multi-device should be an exception. 

Can you verify the single-device version can pass without customizing the threshold? If so, we (cc @Priya2698) should try to understand why the threshold for multi-device has to be larger.

Reply:
One of the failing tests is LayoutTN_NoComms which doesn't involve any collective communication.
The matmul in that test is of size M=1024, N=512, K=256.
On D devices, the local matmul size is M=512/D, N=512, K=256.

The error is not off by a lot. For example, on 2 devices I see a slightly higher error.
Validation error in output 0 on line 145 in file /home/mcowan/Fuser/tests/cpp/test_multidevice_matmul.cpp.
Detected abs error of: 6.86646e-05
absolute tolerance was set to 2.19122e-05
and relative tolerance set to 2.19122e-07

Distributing across 4 results in a bit more error, but still not much:
Validation error in output 0 on line 145 in file /home/mcowan/Fuser/tests/cpp/test_multidevice_matmul.cpp.
Detected abs error of: 7.62939e-05
absolute tolerance was set to 2.19122e-05
and relative tolerance set to 2.19122e-07

Reply:
> The auto-computed threshold by testValidate should have already considered these errors caused by reduction order. I don't see why multi-device should be an exception.

Reduction order can play a role in the higher errors, however torch baselines don't have a way to tune the reduction order. The current method only looks at the reduction size.

For python benchmarking, where we see validation errors, we are considering relaxing tolerances by ~10%. CC: @naoyam 

> Can you verify the single-device version can pass without customizing the threshold? If so, we (cc @Priya2698) should try to understand why the threshold for multi-device has to be larger.

The error constants were generated by executing a large number of sum reductions and noting the max error. These experimentally generated constants were from single-device reductions. This may also be a reason for differing error constants.

Reply:
The reference computation should also be in `double`.

Reply:
> we are considering relaxing tolerances by ~10%

Good to know. But doesn't seem large enough to tolerate this case though (2.19e-5 vs 6.86e-5). 

> The error constants were generated by executing a large number of sum reductions and noting the max error.

Got it -- it's still empirical and not intended to capture the reduction order that triggers the largest difference. 

@Priya2698 what do you recommend are the next steps? Doing nothing will probably lead to more manual threshold customizations like this for multidevice tests.

Reply:
For single-device cases, I had regenerated validation constants in the past (https://github.com/NVIDIA/Fuser/pull/1593, closed unmerged) but it hadn't helped completely resolve the issue.

For multi-device, we should run a similar experiment to see if the experimentally generated constants differ significantly from existing constants with the number of devices (2/4/8). We may need two sets of constants

Reply:
Small update - I've (1) updated the computation to be done in double, then cast back to float and (2) shrunk K to 64. 
Surprisingly, the allreduce and reduce scatter tests are passing.I thought they would have a larger error since the reduction is both local and across devices, but it is the allgather and no communication (local matmul where we either gather shards from different devices or the output stays sharded) are still slightly off. 

Detected abs error of: 1.14441e-05
    absolute tolerance was set to 8.57934e-06
    and relative tolerance set to 8.57934e-08
    
It looks like we would need even more relaxation then 10%, maybe 2x?

Reply:
Yes, this seems like a much larger error. 
It looks like the experimentally generated error constants for single device are not applicable to multi-device.

Reply:
> We may need two sets of constants

This is probably true... the ring allreduce and reduce-scatter (default implementations for these collectives within a node) are known to be less numerically stable. It won't be an issue on 2 or 4 devices since the number of devices is low, but by 8 devices we might need more tolerance.

Reply:
The above error numbers are for how many devices? 
Are you running into errors on single device as well with `double` reference computations?

Reply:
Above errors were for 2 devices. M=1024,N=512,K=64
I also tried a single device test, with double reference computations, and it was failing.

But, M=256,N=64,K=64 is passing for distributed and single device tests. I tested with 2 and 4 devices so looks promising. I guess the large non-reduction axis sizes were increasing the chances of one element being off by too much?